chore: stop using go:linkname for http.registerOnHitEOF, http.requestBodyRemains

hunshcn opened the pull request:
relate to https://github.com/MetaCubeX/mihomo/pull/952#issuecomment-2118639385

hunshcn commented:
按照标准库实现了一下

wwqgtxx commented:
@hunshcn 我比较好奇的是可以覆盖原来requestBodyRemains的判断场景么

hunshcn commented:
> @hunshcn 我比较好奇的是可以覆盖原来requestBodyRemains的判断场景么

```
// requestBodyRemains reports whether future calls to Read
// on rc might yield more data.
func requestBodyRemains(rc io.ReadCloser) bool {
	if rc == NoBody {
		return false
	}
	switch v := rc.(type) {
	case *expectContinueReader:
		return requestBodyRemains(v.readCloser)
	case *body:
		return v.bodyRemains()
	default:
		panic("unexpected type " + fmt.Sprintf("%T", rc))
	}
}
```

由于我们的 request 是 readRequest 读取出来的，通过源码可以得知其实 Body 只有可能是 `*body / NoBody`

所以这个pr中可以直接`if request.Body == nil || request.Body == http.NoBody {`
其实这个 if nil 都可以不需要，但是作为防御性编程。

hunshcn commented:
话说 go.mod 不 stable 啊...是不是得 update 一下并且在 ci 加个 verify go mod

wwqgtxx commented:
> 话说 go.mod 不 stable 啊...是不是得 update 一下并且在 ci 加个 verify go mod

我自己提交的时候都是go mod tidy再提交的，应该没有问题

hunshcn commented:
```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   test/go.mod
	modified:   test/go.sum
```
test/go.mod 不 stable @wwqgtxx

wwqgtxx commented:
test应该是很久没管过了，每次也没有同步更新它的习惯，或者说从clash分支出来之后就从没用过test模块
